### PR TITLE
Building plugin

### PR DESCRIPTION
**Pull request by tobrun**
This PR adds a Building plugin, this plugin API matches Traffic plugin API.

westminster abbey

### CODE REVIEW

**Comment by tobrun** [comment]
I finished the initial stab and this PR is ready for review.

Possible API to add:
 - color?
 - opacity?

**Comment by zugaldia** [comment]
@tobrun Agreed. I think an API for the three different values you set on https://github.com/mapbox/mapbox-plugins-android/pull/26/files#diff-a1a6e9c8cc422ca880b018bba0a5741cR101 (`minZoom`, `color`, `opacity`) is all we need for now.

**Comment by tobrun** [comment]
I have added the API and reworked the code to reflect. This PR is ready for review, though we should maybe wait with merging until v5.1.0 final is released of the Mapbox GL-native dependency?

**Comment by zugaldia** [comment]
> This PR is ready for review, though we should maybe wait with merging until v5.1.0 final is released?

👍 Makes sense to me to wait until the underlying API gets stable and published to release 0.1.0. Meanwhile, I suggest we start publishing SNAPSHOTs of the plugin so that interested devs can start playing with it.

**Comment by tobrun** [comment]
This PR has been updated with latest SDK release and adds Light integration (just a setter for now)

**Comment by cammace** [comment]
> Duplicate import to line 8 - timber.log.Timber. [RedundantImport]

Checkstyle strikes again 😆

File: plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+ */
+public class BuildingActivity extends AppCompatActivity implements OnMapReadyCallback {
+
+  @BindView(R.id.mapView)
+  MapView mapView;
+
+  @BindView(R.id.fabBuilding)
+  View fab;
+
+  @BindView(R.id.seekbar_light_radialCoordinate)
```
Comment (cammace) [inline plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java]: half resources use camel casing, the rest use snake

File: plugins/building/gradle.properties [hunk 2]
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=mapbox-android-building-plugin
```
Comment (cammace) [inline plugins/building/gradle.properties]: To keep consistency around plugins, the artifact id should be `mapbox-android-plugin-building`

File: plugins/building/gradle.properties [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+POM_NAME=Mapbox Android Plugins
+POM_DESCRIPTION=Mapbox Android Plugins (Building)
```
Comment (cammace) [inline plugins/building/gradle.properties]: Mapbox Android Building Plugin

File: plugins/building/proguard-rules.pro [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# Add project specific ProGuard rules here.
```
Comment (cammace) [inline plugins/building/proguard-rules.pro]: file can be deleted?

**Comment by cammace** [comment]
Went ahead and fixed the issues I raised previously. One thing I noticed is that we use `setVisibility(boolean visible)` in this plugin but in traffic we have `toggle()`. Since they essentially do the same thing, we should make the APIs the same (I prefer the `setVisibility` API).

Another addition that should be made in this PR would be the zoom function modifies the building height so they don't just pop in.

**Comment by cammace** [comment]
If this is supported across different styles, we should make sure everything is being redrawn correctly when styles change.

File: plugins/build.gradle [hunk 5]
```diff
@@ -10,6 +10,7 @@ buildscript {
 allprojects {
     repositories {
         jcenter()
+        maven { url "http://oss.sonatype.org/content/repositories/snapshots/" }
```
Comment (zugaldia) [inline plugins/build.gradle]: What's this needed for?

File: plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+  public void onMapReady(MapboxMap mapboxMap) {
+    this.mapboxMap = mapboxMap;
+    buildingPlugin = new BuildingPlugin(mapView, mapboxMap);
+    buildingPlugin.setMinZoomLevel(15);
+    fab.setVisibility(View.VISIBLE);
+    initLightSeekbar();
+  }
+
+  private void initLightSeekbar() {
+    seekbarRadialCoordinate.setMax(24); // unknown?
```
Comment (zugaldia) [inline plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java]: What does `unknown` mean in this context?

File: plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+  private class PositionChangeListener implements SeekBar.OnSeekBarChangeListener {
+    @Override
+    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
+      invalidateLightPosition();
+    }
+
+    @Override
+    public void onStartTrackingTouch(SeekBar seekBar) {
+
```
Comment (zugaldia) [inline plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java]: Could we add a comment or a Timber call in the unused methods to explicitly note they aren't used on purpose?

File: plugins/build.gradle (hunk 5 above)
Comment (cammace) [inline plugins/build.gradle]: leftover from when we were using map SDK snapshot I think. I removed it.

File: plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java (hunk 6 above)
Comment (cammace) [inline plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java]: @tobrun could speak more on this, I'm not sure either.

Comment (tobrun) [inline plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java]: I'm not sure what the maximum radius is for a light source, refs https://en.wikipedia.org/wiki/Spherical_coordinate_system

Comment (zugaldia) [inline plugins/app/src/main/java/com/mapbox/mapboxsdk/plugins/testapp/activity/BuildingActivity.java]: Could we add a note on code mentioning this?

**Pull request merged**